Convert MSC code Cleanup to Formatter

JustinHennis1 opened the pull request:
Closes #13227 

Removed Cleanup functionality for MSC Code conversion. Added formatter at bottom of preset panel because formatters handle single-field conversion. 

### Screenshot Before
![beforeformat6_2_25](https://github.com/user-attachments/assets/4a8b41a2-ef7e-4849-a629-e8c687829f3b)

### Screenshot After
![afterformat6_2_25](https://github.com/user-attachments/assets/908c5ab8-a5ba-4b50-b813-86c9bb89b032)

### Mandatory checks

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (if change is visible to the user)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

koppor commented on CHANGELOG.md:
```diff
@@ -31,6 +31,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 
 ### Changed
 
+- We changed the Cleanup that handled MSC code conversion into a Formatter [#13233](https://github.com/JabRef/jabref/pull/13233)
```
Can be removed - no change since the **last** release

Modify the other changelog entry (if needed) - I didn't check

koppor commented on jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:
```diff
@@ -0,0 +1,89 @@
+import org.slf4j.LoggerFactory;
+
+public class ConvertMSCCodesFormatter extends Formatter implements LayoutFormatter {
+    private static final Logger logger = LoggerFactory.getLogger(ConvertMSCCodesFormatter.class);
```
Sorry, but why a double logger?

Doesn't checkstyle complain here?

Please use `LOGGER` (see line 21) and move it to the top line

JustinHennis1 replied:
Whoops, I overlooked it. Checkstyle passed on this, which is weird.

koppor commented on jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:
```diff
@@ -0,0 +1,89 @@
+            }
+        }
+
+        MSCMAP = tempMap;
```
Assign this before the `if`s

Then you can exit early at each of:

```java
if (resourceUrl == null) {
            logger.error("Resource not found: msc_codes.json");
koppor marked this conversation as resolved.
@trag-bot[Jun 3, 2025](https://github.com/JabRef/jabref/pull/13233/files#r2124518557)

            conversionPossible = false;
           return;
}
try  ..
```

koppor commented on jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:
```diff
@@ -0,0 +1,89 @@
+    public String format(String text) {
+        StringBuilder result = new StringBuilder(Objects.requireNonNull(text));
+
+        if (result.isEmpty() || !conversionPossible) {
+            return result.toString();
+        }
```
This should right as first lines of the method - and use `text`

koppor commented on jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:
```diff
@@ -0,0 +1,89 @@
+            return result.toString();
+        }
+        // text contains comma separated codes
+        String[] codeList = result.toString().split(",");
```
Use `org.jabref.model.entry.KeywordList#parse(java.lang.String, java.lang.Character, java.lang.Character)` and then work on `KeywordList`

Reason: Better data structure and adheres to user configuration (keyword delimiter) and escpaings.

See at other usages of the parser and how to pass the prefernces

---

At first, I thought:

```suggestion
        String[] codeList = text.toString().split(",");
```

but this is wrong, because no escaping treatment etc.

koppor reviewed (changes requested):
Please revet changes in sub modules - see https://devdocs.jabref.org/code-howtos/faq.html#submodules

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests – jablib
-->

koppor commented on jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:
```diff
@@ -0,0 +1,117 @@
+        // get preferences
+        Character dlim = preferences.getKeywordSeparator();
+        Character hdlim = Keyword.DEFAULT_HIERARCHICAL_DELIMITER;
+//        LOGGER.info("Delimiter = {}", dlim);
+//        LOGGER.info("Hierarchical Delimiter = {}", hdlim);
```
You can leave this in with `DEBUG` and increase the log level on your side during development (and do not commit this change)

See https://devdocs.jabref.org/code-howtos/logging.html for details.

koppor commented on jablib/src/main/java/org/jabref/logic/cleanup/FieldFormatterCleanups.java:
```diff
@@ -87,6 +90,13 @@ public FieldFormatterCleanups(boolean enabled, List<FieldFormatterCleanup> actio
         this.actions = Objects.requireNonNull(actions);
     }
 
+    public FieldFormatterCleanups(boolean enabled, List<FieldFormatterCleanup> actions, BibEntryPreferences preferences) {
```
Needs to be reviewed in IDE - why this general method is needed - until now, this was not needed - and the formatter worked.

JustinHennis1 replied:
I had no way to retrieve BibEntryPreferences so that I could call getKeywordSeperator. This is used to inject preferences so that the formatter can use a dynamic delimiter.

If there is another method of retrieving BibEntryPreferences pleases lmk.

JustinHennis1 replied:
Why is getInstance() deprecated for JabrefCLIPreferences Class? Isn't the point of this class to allow preferences to be accessible globally. @koppor

I feel like it defeats the purpose to implement a Singleton design that isn't easily accessible. 😅

JustinHennis1 replied:
Thanks for clarifying. @Siedlerchr 

That said, do you think my approach was correct? 

> Creating a new constructor for FieldFormatterCleanups
> In JabrefCliPreferences, passing BibEntryPreferences to this constructor
> Finally, passing those preferences down to the formatter to get the keyword separator.

koppor replied:
Where is your method used? I did I do a mistake at checking out your branch?

![image](https://github.com/user-attachments/assets/50133eed-9db0-4550-a508-8fa7aa7d3f77)

JustinHennis1 replied:
> Where is your method used? I did I do a mistake at checking out your branch?
> 
> ![image](https://private-user-images.githubusercontent.com/1366654/453770482-50133eed-9db0-4550-a508-8fa7aa7d3f77.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk2MjYwODgsIm5iZiI6MTc0OTYyNTc4OCwicGF0aCI6Ii8xMzY2NjU0LzQ1Mzc3MDQ4Mi01MDEzM2VlZC05ZGIwLTQ1NTAtYTUwOC04ZmE3YWE3ZDNmNzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MTFUMDcwOTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmEyMDMwYTVlNzY3MGIzNTk5YzExOGM0ZjJhZGM3MmQ1M2E3ZDQ2NTIyNzMwMDE0YmFjYjQyOWI1ZGEyY2I5MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.V3KTW7NECFYPjEvjszdxHwcpGfLykV6cVjTJeuoVv4s)

My fault, I think I forgot to add change this (I was testing setter vs constructor injection), I put it in this method of JabRefCliPreferences initially. 
![image](https://github.com/user-attachments/assets/26cc17ed-c88d-438e-9b2c-7e4a97917fbb)

koppor commented:
@JustinHennis1 Could you try to revert the change in the sub module - see https://devdocs.jabref.org/code-howtos/faq.html#submodules

Which git client did you use? Maybe, we can search for a bug report there and ask all developers of JabRef to give a +1 there. (the bug is that `ignore = all` is not handled by the git client)

JustinHennis1 commented:
> @JustinHennis1 Could you try to revert the change in the sub module - see https://devdocs.jabref.org/code-howtos/faq.html#submodules
> 
> Which git client did you use? Maybe, we can search for a bug report there and ask all developers of JabRef to give a +1 there. (the bug is that `ignore = all` is not handled by the git client)

I'm using Git CMD on my Windows Laptop. I just merged main back into my fork, did that fix it?

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests – jablib
-->

JustinHennis1 commented:
@koppor 
### 👋 Hey, I completed the following:
- made changes so that we use Injector to instantiate JabrefCliPreferences in the ConvertMSCCodesFormatter. 
- changed ConvertMSCCodesFormatterTest back to using the default constructor.
- added comments in JabrefCliPreferences specifying why constructor was made public as well as the reason for the changes (without being made public, a java.lang access exception is thrown)
